src/types.ts

```diff
@@ -1,21 +1,29 @@
-export type QueueResult = {
+export type QueueResult<A,B> = {
+  /** Call `fn` with the items in the queue. */
   flush(): Promise<void>
-  enqueue(item: any): Promise<void>
-  lastResult?: any
+  /** Add an item to the queue. When a queue condition is met `flush` will be called. */
+  enqueue(item: A): Promise<void>
+  /** The last result returned from calling `fn`. */
+  lastResult?: Awaited<B>
 }
 
 export interface QueueOptions {
+  /** Wait for the batch to reach this number of elements before flushing the queue. */
   batchSize?: number
+  /** Wait for the batch to reach this size in bytes before flushing the queue. */
   batchBytes?: number
+  /** Wait this long in ms before flushing the queue. */
   timeout?: number
 }
 
-export type Queue = (
-  fn: (arr: any[]) => any,
-  options?: QueueOptions
-) => QueueResult
-
 export interface Deferred {
   done: () => void
   promise: Promise<void>
 }
+
+export interface WaitOptions {
+  /** Wait this long in ms before rejecting. Defaults to 5000 ms. */
+  timeout?: number
+  /** Check the predicate with this frequency. Defaults to 50 ms. */
+  checkFrequency?: number
```

Comment:
Optional suggestion: include the units in the name, so that it's clearer at the call site, e.g. `{ checkFrequencyMs: 100 }`

Reply:
I had that originally but `timeout` already exists for `QueueOptions` and I would rather not introduce inconsistency or a breaking change.